### PR TITLE
Add shared LCOM4 cohesion helper (MethodInfo, cohesion_components)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduces a pure Rust LCOM4 cohesion helper in the common crate, modeling methods as nodes and counting connected components via field accesses and direct method calls.
- Exposes `MethodInfo` and `cohesion_components` from a new `lcom4` module within the common crate.
- Implementation uses an internal Union-Find (path compression + union-by-rank) and deterministic data structures (`BTreeSet<String>`).
- Wired into the crate's public API and covered with unit and behavioural tests. Documentation and roadmap updates accompany the change.

## Changes
- Core implementation
  - Add `common/src/lcom4/mod.rs` with:
    - `MethodInfo` struct: name, accessed_fields (BTreeSet<String>), called_methods (BTreeSet<String>).
    - Private `UnionFind` (path compression + union-by-rank).
    - Helpers to union by shared fields and by method calls.
    - Public `cohesion_components(methods: &[MethodInfo]) -> usize`.
- Public API wiring
  - Update `common/src/lib.rs` to expose the new module:
    - `pub mod lcom4;`
    - `pub use lcom4::{MethodInfo, cohesion_components};`
- Tests
  - Inline unit tests inside `common/src/lcom4/mod.rs` via a separate test module (rstest-based).
  - Behavioural tests:
    - `common/tests/lcom4_behaviour.rs`
    - `common/tests/features/lcom4.feature`
- Documentation
  - Append LCOM4 implementation decisions to:
    - `docs/brain-trust-lints-design.md` (Cohesion analysis section).
  - Roadmap update:
    - Mark 6.1.1 as completed in `docs/roadmap.md`.
  - Execution plan added:
    - `docs/execplans/6-1-1-shared-lcom4-helper.md` with guidance and status.

## Rationale
- Uses a compact union-find approach to count connected components in a graph where nodes are methods and edges exist if:
  - two methods access a common field, or
  - one method calls another on the same type.
- Edge semantics: calls to methods not present in the input are ignored.
- Data is modeled with `BTreeSet<String>` for deterministic iteration and simple API compatibility.
- No dependency on `rustc_private` or HIR types; operates on pre-extracted metadata (`MethodInfo`).

## Tests and validation
- Unit tests cover happy/unhappy/edge cases, including:
  - single method => 1
  - shared field => 1
  - direct/cross calls => 1
  - disjoint methods => >1 components
  - empty input => 0
  - self-calls do not connect distinct methods
- Behavioural tests mirror Gherkin scenarios in `lcom4.feature` via `lcom4_behaviour.rs`.
- All tests run with `cargo test -p common`.

## How to run locally
- Build and test the common crate:
  - cargo test -p common
- Lint and formatting checks (as per project standard):
  - make check-fmt
  - make lint
  - make test

## Acceptance criteria
- [x] `common/src/lcom4/mod.rs` exists and exports `MethodInfo` and `cohesion_components`.
- [x] Inline unit tests cover happy, unhappy, and edge cases.
- [x] Behavioural tests wired for BDD tests under `common/tests`.
- [x] Documentation updated (`brain-trust-lints-design.md`) with implementation decisions.
- [x] Roadmap updated to indicate 6.1.1 as completed.
- [x] Execution plan added (`docs/execplans/6-1-1-shared-lcom4-helper.md`).
- [x] All quality gates (fmt, lint, tests) pass.

---
Notes:
- The module structure uses `common/src/lcom4/mod.rs` (not `lcom4.rs`), and the public API is re-exported from `common/src/lib.rs` as shown above. This aligns with the repository’s module layout and existing patterns.

📎 **Task**: https://www.devboxer.com/task/6c0f1e16-e5af-43b1-8727-c60995b24cf6

### CODE REVIEW

File: common/src/lcom4.rs [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+    }
+}
+
+/// Builds a method-name-to-index map and unions caller/callee pairs.
+fn union_by_method_calls(methods: &[MethodInfo], uf: &mut UnionFind) {
+    let method_index: HashMap<&str, usize> = methods
+        .iter()
+        .enumerate()
+        .map(|(idx, m)| (m.name(), idx))
+        .collect();
```
[the remark this answers is not in the document]
Comment (leynos) [reply common/src/lcom4.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: common/src/lcom4.rs [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+    #[must_use]
+    pub fn accessed_fields(&self) -> &BTreeSet<String> {
+        &self.accessed_fields
+    }
+
+    /// Returns the set of method names called directly by this method.
+    #[must_use]
+    pub fn called_methods(&self) -> &BTreeSet<String> {
+        &self.called_methods
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply common/src/lcom4.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: common/src/lcom4.rs [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+        .iter()
+        .enumerate()
+        .map(|(idx, m)| (m.name(), idx))
+        .collect();
+
+    for (caller_idx, method) in methods.iter().enumerate() {
+        for callee_name in method.called_methods() {
+            if let Some(&callee_idx) = method_index.get(callee_name.as_str()) {
+                uf.union(caller_idx, callee_idx);
+            }
```
[the remark this answers is not in the document]
Comment (leynos) [reply common/src/lcom4.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: common/tests/lcom4_behaviour.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    #[expect(
+        clippy::expect_used,
+        reason = "fixture configuration is required for this behaviour test"
+    )]
+    fn result(&self) -> usize {
+        self.result
+            .get()
+            .expect("LCOM4 result must be computed before assertion")
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply common/tests/lcom4_behaviour.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
